x-pack/legacy/plugins/security/server/lib/authentication/tokens.ts

```diff
@@ -0,0 +1,173 @@
+      // second concurrent request has no idea about that and tries to refresh access token as well. All ends well
+      // when first request refreshes access token and updates session cookie with fresh access/refresh token pair.
+      // But if user navigates to another page _before_ AJAX request (the one that triggered token refresh) responds
+      // with updated cookie, then user will have only that old cookie with expired access token and refresh token
+      // that has been used already.
+      //
+      // When user has neither valid access nor refresh token, the only way to resolve this issue is to get new
+      // SAML LoginResponse and exchange it for a new access/refresh token pair. To do that we initiate a new SAML
+      // handshake. Obviously we can't do that for AJAX requests, so we just reply with `400` and clear error message.
+      // There are two reasons for `400` and not `401`: Elasticsearch search responds with `400` so it seems logical
+      // to do the same on Kibana side and `401` would force user to logout and do full SLO if it's supported.
+      if (getErrorStatusCode(err) === 400) {
+        this.debug('Refresh token is either expired or already used.');
+        return null;
+      }
+
+      throw err;
+    }
+  }
+
+  /**
+   * Tries to invalidate provided access and refresh token pair. At least one of the tokens should
+   * be specified.
+   * @param [accessToken] Optional access token to invalidate.
+   * @param [refreshToken] Optional refresh token to invalidate.
+   */
+  public async invalidate({ accessToken, refreshToken }: Partial<TokenPair>) {
+    this.debug('Invalidating access/refresh token pair.');
+
+    let invalidationError;
+    let invalidatedTokensCount;
+    if (refreshToken) {
+      try {
+        invalidatedTokensCount = (await this.options.client.callWithInternalUser(
+          'shield.deleteAccessToken',
+          { body: { refresh_token: refreshToken } }
+        )).invalidated_tokens;
+      } catch (err) {
+        this.debug(`Failed to invalidate refresh token: ${err.message}`);
+        // We don't re-throw the error here to have a chance to invalidate access token if it's provided.
+        invalidationError = err;
+      }
+
+      if (invalidatedTokensCount === 0) {
+        this.debug('Refresh token was already invalidated.');
+      } else if (invalidatedTokensCount === 1) {
+        this.debug('Refresh token has been successfully invalidated.');
+      } else {
+        this.debug(
+          `${invalidatedTokensCount} refresh tokens were invalidated, this is unexpected.`
+        );
+      }
+    }
+
+    if (accessToken) {
+      try {
+        invalidatedTokensCount = (await this.options.client.callWithInternalUser(
+          'shield.deleteAccessToken',
+          { body: { token: accessToken } }
+        )).invalidated_tokens;
+      } catch (err) {
+        this.debug(`Failed to invalidate access token: ${err.message}`);
+        invalidationError = err;
```

Comment:
Yeah, that's a bit clumsy, I decided to just throw the latest error, but log both. Frankly speaking I can't see any good way of merging these since they can even have different status codes that we return to the client, but I'm all ears if you have any suggestion!

Reply:
Yeah, I don't have any real suggestions and I'm fine with the current approach.